### ISSUE TITLE
Memory leak (on cpu) in 0.9rc (vs. 0.7.1rc)

### ISSUE THREAD

**Issue by osdf** [tensorflow/tensorflow#2728]
Hi,
using the provided wheel vor 0.9rc (also reproduced with a fresh install from source) I'm seeing a lot of memory consumption (on the cpu) during training a model (on a gpu (TITAN X) which gets its data from a ```tf.FIFOQueue``` pinned with ```tf.device()``` to a cpu). I tested two different dataset sizes and these lead to 600MB and 1.2GB leaked memory respectively accumulated during the training loop (GPU memory consumption stays constant during training time).

This memory leak does not happen at all when running the same code using 0.7.1rc. Additionally,
with 0.9rc setting up the whole graph takes about 600MB vs 400MB with 0.7.1rc.

The code is rather large so it will take a bit to get a MWE to (hopefully) reproduce this issue. I suspect it is tied to using a queue (queues, actually, because the validation set is also feed through a different queue, relying on ```tf.template()``` to share model parameters). Has something like that been observed beforehand? Any hints how I could try figure out more of the problem myself?

### Environment info
Operating System:
Ubuntu 15.10

Installed version of CUDA and cuDNN: 
CUDA7.5/cuDNN4.0.4

(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):
``` 
-rw-r--r-- 1 root root 189170 Jan  1 23:25 /usr/local/cuda/lib/libcudadevrt.a
lrwxrwxrwx 1 root root     16 Jan  1 23:25 /usr/local/cuda/lib/libcudart.so -> libcudart.so.7.5
lrwxrwxrwx 1 root root     19 Jan  1 23:25 /usr/local/cuda/lib/libcudart.so.7.5 -> libcudart.so.7.5.18
-rwxr-xr-x 1 root root 311596 Jan  1 23:25 /usr/local/cuda/lib/libcudart.so.7.5.18
-rw-r--r-- 1 root root 558020 Jan  1 23:25 /usr/local/cuda/lib/libcudart_static.a
```
If installed from binary pip package, provide:

1. Which pip package you installed.
https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow-0.9.0rc0-cp27-none-linux_x86_64.whl

2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.
```
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcublas.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcudnn.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcufft.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcuda.so locally
I tensorflow/stream_executor/dso_loader.cc:108] successfully opened CUDA library libcurand.so locally 0.9.0rc0
```

### Steps to reproduce
1. Working on MWE...

### What have you tried?
The described behavior does not happen with 0.7.1rc (UPDATE: also not observed with 0.8rc).

**Comment by neggert** [comment]
I think I'm seeing something similar. Haven't been able to confirm that it's a memory leak, but progress grinds to a halt (`session.run` never returns) after a few iterations of training and memory usage is quite high. I'm also making use of queues, and my setup is almost identical to the one above.

**Comment by jihunchoi** [comment]
I am experiencing the same issue. In python interpreter, `session.run` hangs and it even cannot be stopped until I press Ctrl+\ and terminate the entire process.

**Comment by osdf** [comment]
Updated the original with a bit of more info (training on GPU, GPU memory consumption stays constant). ```session.run``` does not hang for me, training works fine. Though this might be due to the fact that the training set itself is rather small in this case. One minibatch is of size 128x19 (float32).

**Comment by abezuglov** [comment]
Same issue here. Using two Tesla K20c GPU's for training/evaluation. Each training session takes up to a GB of CPU memory which is never returned to the system.

**Comment by nsuke** [comment]
I'm experiencing a similar problem.
When launching a graph, around 16 GB memory is consumed, but only ~3 GB is release upon exiting the process.

Using `top` command, lost 13 GB is only in total used memory, not on any of listed processes.
I.e., summing up every process's memory usage results in a fewer number by ~13 GB.

Memory consumption is constant during training, and re-runing the same graph within a process does not increase the consumption either.
On the other hand, after re-launching another process, the first path consumes another 16 GB.
The memory seems to never go to swap, and when it exhausts the available main memory, Linux hangs.

I'm running the graph on a Titan X on a machine with two Titan X and 64 GB main memory.
I've experienced this for at least master branch at 6/8 and 5/24.
The OS is Ubuntu 16.04.
When I used 0.9rc without GPU support, the leak didn't happen.

Interestingly, when I run very similar graph at home with single GTX 960 machine and 16 GB main memory, this kind of leak never happens.
The batch size fed is much smaller due to GPU memory size, though.
I've tested this on master branch at 6/4 and 6/12.
The OS is ArchLinux.

**Comment by zheng-xq** [comment]
It will be very helpful if someone can post a small case to reproduce the problem.

**Comment by abezuglov** [comment]
I adapted CIFAR multi gpu for my project. In my case **both my code and CIFAR** cause memory leaks on 0.8.0rc version. I rolled back to 0.7.1 and I can confirm that there is no memory leak.
I can provide my code if it is till needed.

**Comment by zheng-xq** [comment]
@abezuglov, does the memory leak only happen with multi-GPU? 

It will be great to have a repro with single GPU. If not, it is okay to have a repro case with multi-GPU.

**Comment by abezuglov** [comment]
If I remember correctly, it did happen with both single and multi GPU's.

Please, find the main training loop below. The complete code is here: [https://github.com/abezuglov/ANN/tree/master/code], where ilt_one_layer.py, ilt_two_layers.py, etc. are the models; ilt_default_feed.py -- training on a single device, ilt_multi_gpu_feed.py -- training on multiple devices. Let me know if I can be of further help.

def train():
    # Read datasets 
    train_dataset, valid_dataset, test_dataset = ld.read_data_sets()

    with tf.Graph().as_default(), tf.device('/cpu:0'):
        x = tf.placeholder(tf.float32, [None, FLAGS.input_vars], name='x-input')
        y_ = tf.placeholder(tf.float32, [None, FLAGS.output_vars], name = 'y-input')

        # Prepare global step and learning rate for optimization...

        optimizer = tf.train.AdamOptimizer(learning_rate)
        outputs = ilt.inference(x)
        loss = ilt.loss(outputs, y_)

        # Calculate gradients and apply them
        grads = optimizer.compute_gradients(loss)
        apply_gradient_op = optimizer.apply_gradients(grads, global_step = global_step)

        # Smoothen variables after gradient applications
        variable_averages = tf.train.ExponentialMovingAverage(
            FLAGS.moving_avg_decay, global_step)
        variables_averages_op = variable_averages.apply(tf.trainable_variables())
        train_op = tf.group(apply_gradient_op, variables_averages_op)

        merged = tf.merge_all_summaries()
           
        init = tf.initialize_all_variables()
        sess = tf.Session(config = tf.ConfigProto(
            allow_soft_placement = False, # allows to utilize GPU's & CPU's
            log_device_placement = False)) # shows GPU/CPU allocation

        # Finish graph creation. Below is the code for running graph
        sess.run(init)
        tf.train.start_queue_runners(sess=sess)

        # Main training loop
        for step in xrange(FLAGS.max_steps):
            start_time = time.time()
            
            _, train_loss, summary, lr = sess.run(
                [train_op, loss, merged, learning_rate], feed_dict=fill_feed_dict(train_dataset, x, y_, train = True))

            duration = time.time()-start_time
            if step%(FLAGS.max_steps//20) == 0:
                feed_dict = fill_feed_dict(valid_dataset, x, y_, train = False)
                valid_loss, summary = sess.run([loss, merged], feed_dict = feed_dict)
                print('Step %d (%.2f op/sec): Training MSE: %.5f, Validation MSE: %.5f' % (
                    step, 1.0/duration, np.float32(train_loss).item(), np.float32(valid_loss).item()))

        feed_dict = fill_feed_dict(test_dataset, x, y_, train = False)
        test_loss = sess.run([loss], feed_dict = feed_dict)
        print('Test MSE: %.5f' % (np.float32(test_loss).item()))
        sess.close()

**Comment by caisq** [comment]
@osdf @neggert @abezuglov @jihunchoi @nsuke Just to gather some info here: What is your NVIDIA driver version (see nvidia-smi output)?

**Comment by abezuglov** [comment]
NVIDIA-SMI 352.93     Driver Version: 352.93

**Comment by ppwwyyxx** [comment]
Ran in to this thread just now. @osdf since you're also suspecting the queue, this might be related to #2942.

**Comment by caisq** [comment]
While @zheng-xq investigates the issue, may I suggest the following? We have seen cases in which similar memory leaks occur for NVIDIA driver 352, but doesn't occur for NVIDIA driver 361. Can you try upgrading the NVIDIA driver and see if the problem persists?

**Comment by nsuke** [comment]
@caisq thanks ! This solved the problem for me.

The leak was happening while using 361.42 from Ubuntu 16.04 official repository.
I've run `apt-get purge nvidia-*` then installed 367.27 from a [PPA](https://launchpad.net/~graphics-drivers/+archive/ubuntu/ppa).
The memory consumption went down from 16+GB to around 6 GB which is now correctly released when existing the process.

**Comment by martinwicke** [comment]
I'm closing this as fixed. Thanks for tracking down the driver issue!
